src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -176,422 +227,317 @@ public void DisconnectFromWorkspace()
-                    return null;
+                // We have potentially heard about several changes to the subject buffer.  However
+                // we only need to process the latest once.
+                Contract.ThrowIfTrue(snapshots.IsDefaultOrEmpty);
+                var currentSnapshot = GetLatest(snapshots);
 
-                var classifiedSpans = ClassificationUtilities.GetOrCreateClassifiedSpanList();
+                var classificationService = TryGetClassificationService(currentSnapshot);
+                if (classificationService == null)
+                    return;
 
-                foreach (var span in spans)
-                {
-                    AddClassifiedSpans(classificationService, span, classifiedSpans);
-                }
+                var currentDocument = currentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (currentDocument == null)
+                    return;
 
-                return ClassificationUtilities.ConvertAndReturnList(
-                    _typeMap, spans[0].Snapshot, classifiedSpans);
-            }
+                GetLastProcessedData(out var previousSnapshot, out var previousDocument, out var previousRoot);
 
-            private void AddClassifiedSpans(
-                IClassificationService classificationService,
-                SnapshotSpan span,
-                List<ClassifiedSpan> classifiedSpans)
-            {
-                // First, get the tree and snapshot that we'll be operating over.  
-                // From this point on we'll do all operations over these values.
-                ITextSnapshot lastSnapshot;
-                Document lastDocument;
-                object lastCachedData;
+                var currentRoot = await currentDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var changedSpan = await ComputeChangedSpanAsync().ConfigureAwait(false);
 
                 lock (_gate)
                 {
-                    lastSnapshot = _lastProcessedSnapshot;
-                    lastDocument = _lastProcessedDocument;
-                    lastCachedData = _lastProcessedCachedData;
+                    _lastProcessedSnapshot = currentSnapshot;
+                    _lastProcessedDocument = currentDocument;
+                    _lastProcessedRoot = currentRoot;
                 }
 
-                if (lastDocument == null)
-                {
-                    // We don't have a syntax tree yet.  Just do a lexical classification of the document.
-                    AddClassifiedSpansForTokens(classificationService, span, classifiedSpans);
-                    return;
-                }
+                // Notify the editor now that there were changes.  Note: we do not need to go the
+                // UI to do this.  The editor supports TagsChanged being called on any thread.
+                this.TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(changedSpan));
```

Comment:
another nicety: we don't have to go back to the ui thread either for notifications.  so the mainline typing case never does anything on the ui thread until the editor calls back into us for GetTags.